tests/e2e/cluster_downgrade_test.go

```diff
@@ -40,30 +48,56 @@ func TestDowngradeUpgrade(t *testing.T) {
 	lastVersionStr := fmt.Sprintf("%d.%d", lastVersion.Major, lastVersion.Minor)
 
 	e2e.BeforeTest(t)
-	dataDirPath := t.TempDir()
 
-	epc := startEtcd(t, currentEtcdBinary, dataDirPath)
-	validateVersion(t, epc, version.Versions{Cluster: currentVersionStr, Server: currentVersionStr})
+	t.Logf("Create cluster with version %s", currentVersionStr)
+	epc := newCluster(t, currentEtcdBinary, clusterSize)
+	for i := 0; i < len(epc.Procs); i++ {
+		validateVersion(t, epc.Cfg, epc.Procs[i], version.Versions{Cluster: currentVersionStr, Server: currentVersionStr})
+	}
+	t.Logf("Cluster created")
 
+	t.Logf("etcdctl downgrade enable %s", lastVersion)
 	downgradeEnable(t, epc, lastVersion)
-	expectLog(t, epc, "The server is ready to downgrade")
-	validateVersion(t, epc, version.Versions{Cluster: lastVersionStr, Server: currentVersionStr})
+	t.Log("Write keys to ensure wal snapshot is created and cluster version set is snapshotted")
+	triggerSnaphsot(t, epc)
 
-	stopEtcd(t, epc)
-	epc = startEtcd(t, lastReleaseBinary, dataDirPath)
-	expectLog(t, epc, "the cluster has been downgraded")
-	validateVersion(t, epc, version.Versions{Cluster: lastVersionStr, Server: lastVersionStr})
+	t.Log("Downgrade enabled, validating if cluster is ready for downgrade")
+	for i := 0; i < len(epc.Procs); i++ {
+		expectLog(t, epc.Procs[i], "The server is ready to downgrade")
+		validateVersion(t, epc.Cfg, epc.Procs[i], version.Versions{Cluster: lastVersionStr, Server: currentVersionStr})
+	}
+	t.Log("Cluster is ready for downgrade")
 
-	stopEtcd(t, epc)
-	epc = startEtcd(t, currentEtcdBinary, dataDirPath)
-	validateVersion(t, epc, version.Versions{Cluster: currentVersionStr, Server: currentVersionStr})
+	t.Log("Starting downgrade process")
+	for i := 0; i < len(epc.Procs); i++ {
+		t.Logf("Downgrading member %d", i)
+		stopEtcd(t, epc.Procs[i])
+		startEtcd(t, epc.Procs[i], lastReleaseBinary)
+	}
+	t.Log("All members downgraded, validating downgrade")
+	expectLog(t, leader(t, epc), "the cluster has been downgraded")
+	for i := 0; i < len(epc.Procs); i++ {
+		validateVersion(t, epc.Cfg, epc.Procs[i], version.Versions{Cluster: lastVersionStr, Server: lastVersionStr})
+	}
+	t.Log("Downgrade complete")
+
+	t.Log("Starting upgrade process")
+	for i := 0; i < len(epc.Procs); i++ {
+		t.Logf("Upgrading member %d", i)
+		stopEtcd(t, epc.Procs[i])
+		startEtcd(t, epc.Procs[i], currentEtcdBinary)
```

Comment:
Shell we validate here that cluster is not upgraded if quorum is not yet upgraded (1-st iteration),
and that cluster gets updated if quorum (including leader ?) get's upgraded ?